% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/connections.R
\name{h_get_granges_by_id}
\alias{h_get_granges_by_id}
\title{Conversion of BioMart Coordinates into \code{GRanges}}
\usage{
h_get_granges_by_id(df, id)
}
\arguments{
\item{df}{(\code{data.frame})\cr data.frame as returned by \code{biomaRt::getBM()}.}

\item{id}{(\code{character})\cr gene ID(s) to convert the coordinates for.}
}
\value{
\code{GRange} objects for the respective gene ID(s).
}
\description{
\ifelse{html}{\href{https://lifecycle.r-lib.org/articles/stages.html#experimental}{\figure{lifecycle-experimental.svg}{options: alt='[Experimental]'}}}{\strong{[Experimental]}}

This function automatically extracts the chromosome number, the start position and the end position of transcripts
in given data.frame as returned by \code{biomaRt::getBM()} and converts them to \code{GRanges} objects.
}
\examples{
if(interactive()){
  mart <- biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  attrs <- c("ensembl_gene_id",
             "ensembl_exon_id",
             "chromosome_name",
             "exon_chrom_start",
             "exon_chrom_end")
  coords <- biomaRt::getBM(
    filters = "entrezgene_id",
    attributes = attrs,
    values = c("11185", "10677"),
    mart = mart
  )
  h_get_granges_by_id(coords, "ENSG00000135407")
}
}
