---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

[![Project Status: Active â€“ The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
\ 

# hermes <p align="center"> <img src='man/figures/logo.png' align="right" height="131.5" alt="hermes-logo"/> </p>

> `hermes` facilitates preprocessing, analyzing, and reporting of RNA-seq data.

* Imports RNAseq count data into the `hermes` ready format.
* Annotates gene information automatically from a central database (e.g. BioMart).
* Adds quality control (QC) flags to genes and samples.
* Filters the data set.
* Normalizes the counts.
* Quickly produces descriptive plots.
* Performs principal components analysis.
* Produces a templated QC report.
* Performs differential expression analysis.

## Installation

You can install the development version from GitHub with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
devtools::install_github("insightsengineering/hermes")
```

## Getting Started

You can get started by reading the introduction vignette:

```{r vignette-start, eval = FALSE}
library(hermes)
vignette("introduction", package = "hermes")
```
