<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • hermes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://insightsengineering.github.io/hermes/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-125641273-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-125641273-1');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hermes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/introduction.html">Introduction to `hermes`</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/insightsengineering/hermes" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/insightsengineering/hermes/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.5" id="hermes-0995">hermes 0.99.5<a class="anchor" aria-label="anchor" href="#hermes-0995"></a></h2>
<div class="section level4">
<h4 id="meta-0-99-5">Meta<a class="anchor" aria-label="anchor" href="#meta-0-99-5"></a></h4>
<ul><li>First public release of the <code>hermes</code> package.</li>
<li>Submission to BioConductor.</li>
</ul></div>
<div class="section level4">
<h4 id="enhancements-0-99-5">Enhancements<a class="anchor" aria-label="anchor" href="#enhancements-0-99-5"></a></h4>
<ul><li>Better legends on the genes barplot and the correlation heatmap.</li>
<li>Improved vignette layout using the BioConductor style.</li>
<li>Additional <code>version</code> argument for <code>connect_biomart</code> to specify an Ensembl version.</li>
</ul></div>
<div class="section level4">
<h4 id="miscellaneous-0-99-5">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous-0-99-5"></a></h4>
<ul><li>New utility function <code>cut_quantile</code> for cutting a numeric vector into quantiles.</li>
<li>New utility function <code>cat_with_newline</code> for concatenating and printing with newline.</li>
<li>New check function <code>check_proportion</code> which checks for a single proportion.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1" id="hermes-011">hermes 0.1.1<a class="anchor" aria-label="anchor" href="#hermes-011"></a></h2>
<div class="section level4">
<h4 id="new-features-0-1-1">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-1-1"></a></h4>
<ul><li>New function <code>draw_scatterplot</code> to produce scatterplots of two genes or gene signatures.</li>
<li>New function <code>draw_boxplot</code> for boxplots of gene expression values.</li>
<li>New function <code>draw_barplot</code> for barplots of dichotomized gene expression counts into two or three percentile categories.</li>
<li>New helper function <code>wrap_in_mae</code> that wraps a single <code>SummarizedExperiment</code> object into an MAE object.</li>
<li>New method <code>rename</code> that makes renaming columns of <code>rowData</code> and <code>colData</code> as well as assay names in existing <code>SummarizedExperiment</code> objects much easier, as a step before converting to <code>HermesData</code>.</li>
<li>New method <code>lapply</code> that allows user to apply a function on all experiments in a <code>MultiAssayExperiment</code>.</li>
<li>New method <code>isEmpty</code> that checks whether a <code>SummarizedExperiment</code> object is empty.</li>
<li>New gene filtering option <code>n_top</code> in the <code>calc_pca</code> function, which allows filtering genes with greatest variability across samples.</li>
<li>New class <code>GeneSpec</code> for specification of genes or gene signatures, see <code><a href="../reference/gene_spec.html">?gene_spec</a></code> for simple construction. Inclusion of gene signature functions <code>colPrinComp1</code> and <code>colMeanZscores</code> to supplement standard column statistics functions.</li>
<li>New helper function <code>col_data_with_genes</code> which extracts the sample variables saved in <code>colData</code> together with selected gene information as a combined data set.</li>
<li>New helper function <code>inner_join_cdisc</code> which joins genetic with CDISC data sets.</li>
</ul></div>
<div class="section level4">
<h4 id="bug-fixes-0-1-1">Bug Fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-1-1"></a></h4>
<ul><li>
<code><a href="../reference/normalize.html">normalize()</a></code> now also works when the <code>hermes</code> package is not loaded, i.e. you can use it with <code><a href="../reference/normalize.html">hermes::normalize()</a></code>.</li>
<li>
<code><a href="../reference/correlate.html">correlate()</a></code> now also works when there are factor variables in the sample variables of the <code>HermesData</code> object.</li>
<li>
<code><a href="../reference/quality_flags.html">add_quality_flags()</a></code> does no longer return <code>NA</code> as the technical failure flags for the samples if there is only a single gene contained in the input, but instead a vector of <code>FALSE</code> to ensure correct downstream functionality.</li>
</ul></div>
<div class="section level4">
<h4 id="miscellaneous-0-1-1">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous-0-1-1"></a></h4>
<ul><li>Updated <code>LICENCE</code> and <code>README</code> with new package references.</li>
<li>The <code>multi_assay_experiment</code> now contains <code>HermesData</code> experiments, different patient IDs, one experiment with normalized assays, and multiple samples per patient in one experiment.</li>
<li>The main <code>HermesData</code> example is now saved in the package as <code>hermes_data</code>, and the previous <code>summarized_experiment</code> is still available. Note that patient IDs have been changed in the new version to align with the <code>multi_assay_experiment</code>.</li>
<li>Renaming of required <code>rowData</code> and <code>colData</code> columns to be more consistent with standards and use lowercase snake-case names.</li>
<li>Annotation querying and setting is now more flexible in that it also allows to query more annotations than the required ones.</li>
<li>Instead of gene starts and ends, the total length of gene exons is now used as the annotation column <code>size</code>. Corresponding queries from BioMart are used to return this gene size.</li>
<li>
<code>df_char_to_factor</code> has been deprecated (and can still be used with a warning) and replaced with <code>df_cols_to_factor</code>, which also converts logical variables to factor variables.</li>
<li>When providing <code>SummarizedExperiment</code> objects containing <code>DelayedMatrix</code> assays to the <code><a href="../reference/HermesData-class.html">HermesData()</a></code> constructor, these are silently converted to <code>matrix</code> assays to ensure downstream functionality.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.0" id="hermes-010">hermes 0.1.0<a class="anchor" aria-label="anchor" href="#hermes-010"></a></h2>
<ul><li>First internal release of the <code>hermes</code> package, which contains classes, methods and functions to import, quality-check, filter, normalize, and analyze RNAseq counts data for differential expression.</li>
<li>
<code>hermes</code> is a successor of the <code>rnaseqTools</code> R package. The core functionality is built on the BioConductor ecosystem, especially the <code>SummarizedExperiment</code> class. New users should first begin by reading the “Introduction to <code>hermes</code>” vignette to become familiar with the <code>hermes</code> concepts.</li>
</ul><div class="section level4">
<h4 id="new-features-0-1-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-1-0"></a></h4>
<ul><li>Import RNAseq count data into the <code>hermes</code> ready format.</li>
<li>Annotate gene information from the Ensembl database via <code>biomaRt</code>.</li>
<li>Add quality control (QC) flags to genes and samples.</li>
<li>Filter and subset the data set.</li>
<li>Normalize the counts.</li>
<li>Produce descriptive plots.</li>
<li>Perform principal components analysis.</li>
<li>Produce a templated QC Rmd report.</li>
<li>Perform differential expression analysis.</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Daniel Sabanés Bové, Namrata Bhatia, Stefanie Bienert, Benoit Falquet, Haocheng Li, Jeff Luong, Lyndsee Midori Zhang, Simona Rossomanno, Tim Treis, Mark Yan, Naomi Chang, Chendi Liao, Carolyn Zhang, Joseph N. Paulson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

