<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Normalization of AnyHermesData Objects — normalize,AnyHermesData-method • hermes</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Normalization of AnyHermesData Objects — normalize,AnyHermesData-method"><meta name="description" content="
The normalize() method is normalizing the input AnyHermesData according to one or more
specified normalization methods. The results are saved as additional assays
in the object.
Possible normalization methods (which are implemented with separate helper functions):
cpm: Counts per Million (CPM). Separately by sample, the original counts of the genes
are divided by the library size of this sample, and multiplied by one million. This is the
appropriate normalization for between-sample comparisons.
rpkm: Reads per Kilobase of transcript per Million reads mapped (RPKM). Each gene count is
divided by the gene size (in kilobases) and then again divided by the library sizes of each
sample (in millions). This allows for within-sample comparisons, as it takes
into account the gene sizes - longer genes will always have more counts than shorter genes.
tpm: Transcripts per Million (TPM). This addresses the problem of RPKM being inconsistent
across samples (which can be seen that the sum of all RPKM values will vary from sample to
sample). Therefore here we divide the RPKM by the sum of all RPKM values for each sample,
and multiply by one million.
voom: VOOM normalization. This is essentially just a slight variation of CPM where
a prior_count of 0.5 is combined with lib_sizes increased by 1 for each sample. Note that
this is not required for the corresponding differential expression analysis, but just provided
as a complementary experimental normalization approach here.
vst: Variance stabilizing transformation. This is to transform the normalized
count data for all genes into approximately homoskedastic values (having constant variance).
rlog: The transformation to the log2 scale values with approximately homoskedastic values.

"><meta property="og:description" content="
The normalize() method is normalizing the input AnyHermesData according to one or more
specified normalization methods. The results are saved as additional assays
in the object.
Possible normalization methods (which are implemented with separate helper functions):
cpm: Counts per Million (CPM). Separately by sample, the original counts of the genes
are divided by the library size of this sample, and multiplied by one million. This is the
appropriate normalization for between-sample comparisons.
rpkm: Reads per Kilobase of transcript per Million reads mapped (RPKM). Each gene count is
divided by the gene size (in kilobases) and then again divided by the library sizes of each
sample (in millions). This allows for within-sample comparisons, as it takes
into account the gene sizes - longer genes will always have more counts than shorter genes.
tpm: Transcripts per Million (TPM). This addresses the problem of RPKM being inconsistent
across samples (which can be seen that the sum of all RPKM values will vary from sample to
sample). Therefore here we divide the RPKM by the sum of all RPKM values for each sample,
and multiply by one million.
voom: VOOM normalization. This is essentially just a slight variation of CPM where
a prior_count of 0.5 is combined with lib_sizes increased by 1 for each sample. Note that
this is not required for the corresponding differential expression analysis, but just provided
as a complementary experimental normalization approach here.
vst: Variance stabilizing transformation. This is to transform the normalized
count data for all genes into approximately homoskedastic values (having constant variance).
rlog: The transformation to the log2 scale values with approximately homoskedastic values.

"><meta property="og:image" content="https://insightsengineering.github.io/hermes/logo.png"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-125641273-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-125641273-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hermes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.2.9004</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/hermes.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/insightsengineering/hermes"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Normalization of <code>AnyHermesData</code> Objects</h1>
      <small class="dont-index">Source: <a href="https://github.com/insightsengineering/hermes/blob/main/R/normalization.R" class="external-link"><code>R/normalization.R</code></a></small>
      <div class="d-none name"><code>normalize.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="figures/lifecycle-stable.svg" alt="[Stable]"></a></p>
<p>The <code>normalize()</code> method is normalizing the input <code><a href="HermesData-class.html">AnyHermesData</a></code> according to one or more
specified normalization methods. The results are saved as additional assays
in the object.</p>
<p>Possible normalization methods (which are implemented with separate helper functions):</p><ul><li><p>cpm: Counts per Million (<code>CPM</code>). Separately by sample, the original counts of the genes
are divided by the library size of this sample, and multiplied by one million. This is the
appropriate normalization for between-sample comparisons.</p></li>
<li><p>rpkm: Reads per Kilobase of transcript per Million reads mapped (RPKM). Each gene count is
divided by the gene size (in kilobases) and then again divided by the library sizes of each
sample (in millions). This allows for within-sample comparisons, as it takes
into account the gene sizes - longer genes will always have more counts than shorter genes.</p></li>
<li><p>tpm: Transcripts per Million (TPM). This addresses the problem of RPKM being inconsistent
across samples (which can be seen that the sum of all RPKM values will vary from sample to
sample). Therefore here we divide the RPKM by the sum of all RPKM values for each sample,
and multiply by one million.</p></li>
<li><p>voom: VOOM normalization. This is essentially just a slight variation of <code>CPM</code> where
a <code>prior_count</code> of 0.5 is combined with <code>lib_sizes</code> increased by 1 for each sample. Note that
this is not required for the corresponding differential expression analysis, but just provided
as a complementary experimental normalization approach here.</p></li>
<li><p><code>vst</code>: Variance stabilizing transformation. This is to transform the normalized
count data for all genes into approximately homoskedastic values (having constant variance).</p></li>
<li><p><code>rlog</code>: The transformation to the log2 scale values with approximately homoskedastic values.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for class 'AnyHermesData'</span></span>
<span><span class="fu">normalize</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cpm"</span>, <span class="st">"rpkm"</span>, <span class="st">"tpm"</span>, <span class="st">"voom"</span>, <span class="st">"vst"</span><span class="op">)</span>,</span>
<span>  control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">h_cpm</span><span class="op">(</span><span class="va">object</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">h_rpkm</span><span class="op">(</span><span class="va">object</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">h_tpm</span><span class="op">(</span><span class="va">object</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">h_voom</span><span class="op">(</span><span class="va">object</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">h_vst</span><span class="op">(</span><span class="va">object</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">h_rlog</span><span class="op">(</span><span class="va">object</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>(<code>AnyHermesData</code>)<br> object to normalize.</p></dd>


<dt id="arg-methods">methods<a class="anchor" aria-label="anchor" href="#arg-methods"></a></dt>
<dd><p>(<code>character</code>)<br> which normalization methods to use, see details.</p></dd>


<dt id="arg-control">control<a class="anchor" aria-label="anchor" href="#arg-control"></a></dt>
<dd><p>(named <code>list</code>)<br> settings produced by <code><a href="control_normalize.html">control_normalize()</a></code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>not used.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The <code><a href="HermesData-class.html">AnyHermesData</a></code> object with additional assays containing the normalized counts.
The <code>control</code> is saved in the <code>metadata</code> of the object for future reference.</p>
    </div>
    <div class="section level2">
    <h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a></h2>

<ul><li><p><code>h_cpm()</code>: calculates the Counts per Million (<code>CPM</code>) normalized counts.</p></li>
<li><p><code>h_rpkm()</code>: calculates the Reads per Kilobase per Million (RPKM) normalized counts.</p></li>
<li><p><code>h_tpm()</code>: calculates the Transcripts per Million (TPM) normalized counts.</p></li>
<li><p><code>h_voom()</code>: calculates the VOOM normalized counts. <a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a></p></li>
<li><p><code>h_vst()</code>: variance stabilizing transformation (<code>vst</code>) from <code>DESeq2</code> package.</p></li>
<li><p><code>h_rlog()</code>: regularized log transformation (<code>rlog</code>) from <code>DESeq2</code> package.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="control_normalize.html">control_normalize()</a></code> to define the normalization method settings.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="va">hermes_data</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># By default, log values are used with a prior count of 1 added to original counts.</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">normalize</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "counts" "cpm"    "rpkm"   "tpm"    "voom"   "vst"   </span>
<span class="r-in"><span><span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">result</span>, <span class="st">"tpm"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tpm</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  06520011B0023R 06520067C0018R 06520063C0043R</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GeneID:11185          0.8626554       3.993877       3.591742</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GeneID:10677          7.0670966       4.054655       3.289554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GeneID:101928428      0.0000000       0.000000       0.000000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># We can also work on original scale.</span></span></span>
<span class="r-in"><span><span class="va">result_orig</span> <span class="op">&lt;-</span> <span class="fu">normalize</span><span class="op">(</span><span class="va">a</span>, control <span class="op">=</span> <span class="fu"><a href="control_normalize.html">control_normalize</a></span><span class="op">(</span>log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tpm_orig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">result_orig</span>, <span class="st">"tpm"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tpm_orig</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  06520011B0023R 06520067C0018R 06520063C0043R</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GeneID:11185          0.8183821       14.93224       11.05653</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GeneID:10677        133.0936048       15.61778        8.77810</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GeneID:101928428      0.0000000        0.00000        0.00000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Separate calculation of the CPM normalized counts.</span></span></span>
<span class="r-in"><span><span class="va">counts_cpm</span> <span class="op">&lt;-</span> <span class="fu">h_cpm</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">counts_cpm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  num [1:5085, 1:20] -0.453 8.252 -2.453 -2.453 -2.453 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Separate calculation of the RPKM normalized counts.</span></span></span>
<span class="r-in"><span><span class="va">counts_rpkm</span> <span class="op">&lt;-</span> <span class="fu">h_rpkm</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">counts_rpkm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  num [1:5085, 1:20] -2.904 4.027 0.404 1.12 1.323 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Separate calculation of the TPM normalized counts.</span></span></span>
<span class="r-in"><span><span class="va">counts_tpm</span> <span class="op">&lt;-</span> <span class="fu">h_tpm</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">counts_tpm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  num [1:5085, 1:20] 0.863 7.067 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Separate calculation of the VOOM normalized counts.</span></span></span>
<span class="r-in"><span><span class="va">counts_voom</span> <span class="op">&lt;-</span> <span class="fu">h_voom</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">counts_voom</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  num [1:5085, 1:20] -0.645 8.251 -3.453 -3.453 -3.453 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Separate calculation of the vst transformation.</span></span></span>
<span class="r-in"><span><span class="va">counts_vst</span> <span class="op">&lt;-</span> <span class="fu">h_vst</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">counts_vst</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  num [1:5085, 1:20] 3.03 10.6 1.78 1.78 1.78 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Separate calculation of the rlog transformation.</span></span></span>
<span class="r-in"><span><span class="va">counts_rlog</span> <span class="op">&lt;-</span> <span class="fu">h_rlog</span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">counts_rlog</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  num [1:5085, 1:20] 2.97 10.13 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "betaPriorVar")= num 7.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "intercept")= num [1:5085, 1] 5.26 8.49 -Inf -Inf -Inf ...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel Sabanés Bové, Namrata Bhatia, Stefanie Bienert, Benoit Falquet, Haocheng Li, Jeff Luong, Lyndsee Midori Zhang, Alex Richardson, Simona Rossomanno, Tim Treis, Mark Yan, Naomi Chang, Chendi Liao, Carolyn Zhang, Joseph N. Paulson, F. Hoffmann-La Roche AG.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

