<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Normalization of AnyHermesData Objects — normalize,AnyHermesData-method • hermes</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Normalization of AnyHermesData Objects — normalize,AnyHermesData-method" />
<meta property="og:description" content="
The normalize() method is normalizing the input AnyHermesData according to one or more
specified normalization methods. The results are saved as additional assays
in the object.
Possible normalization methods (which are implemented with separate helper functions):
cpm: Counts per Million (CPM). Separately by sample, the original counts of the genes
are divided by the library size of this sample, and multiplied by one million. This is the
appropriate normalization for between-sample comparisons.
rpkm: Reads per Kilobase of transcript per Million reads mapped (RPKM). Each gene count is
divided by the gene size (in kilobases) and then again divided by the library sizes of each
sample (in millions). This allows for within-sample comparisons, as it takes
into account the gene sizes - longer genes will always have more counts than shorter genes.
tpm: Transcripts per Million (TPM). This addresses the problem of RPKM being inconsistent
across samples (which can be seen that the sum of all RPKM values will vary from sample to
sample). Therefore here we divide the RPKM by the sum of all RPKM values for each sample,
and multiply by one million.
voom: VOOM normalization. This is essentially just a slight variation of CPM where
a prior_count of 0.5 is combined with lib_sizes increased by 1 for each sample. Note that
this is not required for the corresponding differential expression analysis, but just provided
as a complementary experimental normalization approach here.
vst: Variance stabilizing transformation. This is to transform the normalized
count data for all genes into approximately homoskedastic values (having constant variance).
rlog: The transformation to the log2 scale values with approximately homoskedastic values.

" />
<meta property="og:image" content="https://insightsengineering.github.io/hermes/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-125641273-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-125641273-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hermes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">Introduction to `hermes`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/insightsengineering/hermes">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Normalization of <code>AnyHermesData</code> Objects</h1>
    <small class="dont-index">Source: <a href='https://github.com/insightsengineering/hermes/blob/master/R/normalization.R'><code>R/normalization.R</code></a></small>
    <div class="hidden name"><code>normalize.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><a href='https://lifecycle.r-lib.org/articles/stages.html#stable'><img src='figures/lifecycle-stable.svg' alt='[Stable]' /></a></p>
<p>The <code>normalize()</code> method is normalizing the input <code><a href='HermesData-class.html'>AnyHermesData</a></code> according to one or more
specified normalization methods. The results are saved as additional assays
in the object.</p>
<p>Possible normalization methods (which are implemented with separate helper functions):</p><ul>
<li><p><code>cpm</code>: Counts per Million (CPM). Separately by sample, the original counts of the genes
are divided by the library size of this sample, and multiplied by one million. This is the
appropriate normalization for between-sample comparisons.</p></li>
<li><p><code>rpkm</code>: Reads per Kilobase of transcript per Million reads mapped (RPKM). Each gene count is
divided by the gene size (in kilobases) and then again divided by the library sizes of each
sample (in millions). This allows for within-sample comparisons, as it takes
into account the gene sizes - longer genes will always have more counts than shorter genes.</p></li>
<li><p><code>tpm</code>: Transcripts per Million (TPM). This addresses the problem of RPKM being inconsistent
across samples (which can be seen that the sum of all RPKM values will vary from sample to
sample). Therefore here we divide the RPKM by the sum of all RPKM values for each sample,
and multiply by one million.</p></li>
<li><p><code>voom</code>: VOOM normalization. This is essentially just a slight variation of CPM where
a <code>prior_count</code> of 0.5 is combined with <code>lib_sizes</code> increased by 1 for each sample. Note that
this is not required for the corresponding differential expression analysis, but just provided
as a complementary experimental normalization approach here.</p></li>
<li><p><code>vst</code>: Variance stabilizing transformation. This is to transform the normalized
count data for all genes into approximately homoskedastic values (having constant variance).</p></li>
<li><p><code>rlog</code>: The transformation to the log2 scale values with approximately homoskedastic values.</p></li>
</ul>

    </div>

    <pre class="usage"><span class='co'># S4 method for AnyHermesData</span>
<span class='fu'>normalize</span><span class='op'>(</span>
  <span class='va'>object</span>,
  methods <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"cpm"</span>, <span class='st'>"rpkm"</span>, <span class='st'>"tpm"</span>, <span class='st'>"voom"</span>, <span class='st'>"vst"</span><span class='op'>)</span>,
  control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='fu'>h_cpm</span><span class='op'>(</span><span class='va'>object</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>h_rpkm</span><span class='op'>(</span><span class='va'>object</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>h_tpm</span><span class='op'>(</span><span class='va'>object</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>h_voom</span><span class='op'>(</span><span class='va'>object</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>h_vst</span><span class='op'>(</span><span class='va'>object</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>h_rlog</span><span class='op'>(</span><span class='va'>object</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>(<code>AnyHermesData</code>)<br /> object to normalize.</p></td>
    </tr>
    <tr>
      <th>methods</th>
      <td><p>(<code>character</code>)<br /> which normalization methods to use, see details.</p></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>(named <code>list</code>)<br /> settings produced by <code><a href='control_normalize.html'>control_normalize()</a></code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>not used.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The <code><a href='HermesData-class.html'>AnyHermesData</a></code> object with additional assays containing the normalized counts.
The <code>control</code> is saved in the <code>metadata</code> of the object for future reference.</p>
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    
<ul>
<li><p><code>h_cpm</code>: calculates the Counts per Million (CPM) normalized counts.</p></li>
<li><p><code>h_rpkm</code>: calculates the Reads per Kilobase per Million (RPKM) normalized counts.</p></li>
<li><p><code>h_tpm</code>: calculates the Transcripts per Million (TPM) normalized counts.</p></li>
<li><p><code>h_voom</code>: calculates the VOOM normalized counts. <a href='https://lifecycle.r-lib.org/articles/stages.html#experimental'><img src='figures/lifecycle-experimental.svg' alt='[Experimental]' /></a></p></li>
<li><p><code>h_vst</code>: variance stabilizing transformation (<code>vst</code>) from <code>DESeq2</code> package.</p></li>
<li><p><code>h_rlog</code>: regularized log transformation (<code>rlog</code>) from <code>DESeq2</code> package.</p></li>
</ul>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='control_normalize.html'>control_normalize()</a></code> to define the normalization method settings.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>a</span> <span class='op'>&lt;-</span> <span class='va'>hermes_data</span>

<span class='co'># By default, log values are used with a prior count of 1 added to original counts.</span>
<span class='va'>result</span> <span class='op'>&lt;-</span> <span class='fu'>normalize</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'>assayNames</span><span class='op'>(</span><span class='va'>result</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "counts" "cpm"    "rpkm"   "tpm"    "voom"   "vst"   </div><div class='input'><span class='va'>tpm</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span><span class='va'>result</span>, <span class='st'>"tpm"</span><span class='op'>)</span>
<span class='va'>tpm</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span><span class='op'>]</span>
</div><div class='output co'>#&gt;                  06520011B0023R 06520067C0018R 06520063C0043R
#&gt; GeneID:11185          0.8626554       3.993877       3.591742
#&gt; GeneID:10677          7.0670966       4.054655       3.289554
#&gt; GeneID:101928428      0.0000000       0.000000       0.000000</div><div class='input'>
<span class='co'># We can also work on original scale.</span>
<span class='va'>result_orig</span> <span class='op'>&lt;-</span> <span class='fu'>normalize</span><span class='op'>(</span><span class='va'>a</span>, control <span class='op'>=</span> <span class='fu'><a href='control_normalize.html'>control_normalize</a></span><span class='op'>(</span>log <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>tpm_orig</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span><span class='va'>result_orig</span>, <span class='st'>"tpm"</span><span class='op'>)</span>
<span class='va'>tpm_orig</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span><span class='op'>]</span>
</div><div class='output co'>#&gt;                  06520011B0023R 06520067C0018R 06520063C0043R
#&gt; GeneID:11185          0.8183821       14.93224       11.05653
#&gt; GeneID:10677        133.0936048       15.61778        8.77810
#&gt; GeneID:101928428      0.0000000        0.00000        0.00000</div><div class='input'>
<span class='co'># Separate calculation of the CPM normalized counts.</span>
<span class='va'>counts_cpm</span> <span class='op'>&lt;-</span> <span class='fu'>h_cpm</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>counts_cpm</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  num [1:5085, 1:20] -0.453 8.252 -2.453 -2.453 -2.453 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...
#&gt;   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</div><div class='input'>
<span class='co'># Separate calculation of the RPKM normalized counts.</span>
<span class='va'>counts_rpkm</span> <span class='op'>&lt;-</span> <span class='fu'>h_rpkm</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>counts_rpkm</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  num [1:5085, 1:20] -2.904 4.027 0.404 1.12 1.323 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...
#&gt;   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</div><div class='input'>
<span class='co'># Separate calculation of the TPM normalized counts.</span>
<span class='va'>counts_tpm</span> <span class='op'>&lt;-</span> <span class='fu'>h_tpm</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>counts_tpm</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  num [1:5085, 1:20] 0.863 7.067 0 0 0 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...
#&gt;   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</div><div class='input'>
<span class='co'># Separate calculation of the VOOM normalized counts.</span>
<span class='va'>counts_voom</span> <span class='op'>&lt;-</span> <span class='fu'>h_voom</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>counts_voom</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  num [1:5085, 1:20] -0.645 8.251 -3.453 -3.453 -3.453 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...
#&gt;   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</div><div class='input'>
<span class='co'># Separate calculation of the vst transformation.</span>
<span class='va'>counts_vst</span> <span class='op'>&lt;-</span> <span class='fu'>h_vst</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>counts_vst</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  num [1:5085, 1:20] 3.03 10.6 1.78 1.78 1.78 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...
#&gt;   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...</div><div class='input'>
<span class='co'># Separate calculation of the rlog transformation.</span>
<span class='va'>counts_rlog</span> <span class='op'>&lt;-</span> <span class='fu'>h_rlog</span><span class='op'>(</span><span class='va'>a</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>counts_rlog</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  num [1:5085, 1:20] 2.97 10.13 0 0 0 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:5085] "GeneID:11185" "GeneID:10677" "GeneID:101928428" "GeneID:100422835" ...
#&gt;   ..$ : chr [1:20] "06520011B0023R" "06520067C0018R" "06520063C0043R" "06520105C0017R" ...
#&gt;  - attr(*, "betaPriorVar")= num 7.01
#&gt;  - attr(*, "intercept")= num [1:5085, 1] 5.26 8.49 -Inf -Inf -Inf ...</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Daniel Sabanés Bové, Namrata Bhatia, Stefanie Bienert, Benoit Falquet, Haocheng Li, Jeff Luong, Lyndsee Midori Zhang, Simona Rossomanno.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


