---
title: "Design: Transformation of Counts"
author: "Haocheng Li"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

This vignette explores the design of the transformation workflow and functions in hermes.

## Problems to solve

-  Original `normalize()` function works for the normalization of the data our `HermesData` class. However, the raw counts data after regular normalization procedure (e.g. CPM, TPM, ...) may not be proper for certain analyses (e.g. differential expression analysis, clustering analysis, and discriminant analysis) due to Heteroscedasticity problems (i.e. different variance levels across genes).  
-  To address this issue, one approach is to transform the normalized count data for all genes into approximately homoskedastic values (having constant variance). 
-  We will introduce two transformations to the count data. The linear transformation approach is implemented by a `variance stabilizing transformation` (VST), see [article](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106) for background. The transformation to the log2 scale values is implemented by a 'regularized log' transformation (rlog), see [article](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) for more details.
-  The returned outputs is NOT for differential expression analysis, since the differential expression analysis will be implemented in other `HermesData` functions. The outputs with transformed values are only for visualization or other analyses such as clustering or discriminant analysis.  


## Plan
- Both `vst` and `rlog` transformation methods are available in `DESeq2` package, which are `varianceStabilizingTransformation` and `rlog` functions, respectively. Our functions will call the `DESeq2` package and its functions. 
- We will add two options  `vst` and `rlog` into the `normalize()` function. 
- The `varianceStabilizingTransformation` and `rlog` functions only require a matrix of counts for computation. We could pass the count matrix obtained from `counts()`. 
- Results should be returned as an assay in `HermesData` object as other normalization method.


## Control function

- Add `vst` and `rlog` into the `normalize()` function. 

```{r}
  normalize_update = function(object,
                        methods = c("vst", "rlog"),
                        control = control_normalize(),
                        ...) {
    method_choices <- c( "vst", "rlog")
    assert_that(all(methods %in% method_choices))
    methods <- match.arg(methods, choices = method_choices, several.ok = TRUE)
    for (method in methods) {
      fun_name <- paste0("h_", method)
      method_result <- eval(utils.nest::call_with_colon(
        fun_name,
        object = object,
        control = control
      ))
      assay(object, method) <- method_result
    }
    metadata(object) <- c(metadata(object), list(control_normalize = control))
    object
  }
```

## Helper functions

```{r}
h_vst <- function(object, ...) {
  assert_that(
    is_hermes_data(object)
  )
  DESeq2::varianceStabilizingTransformation(counts(object))
} 
 
h_rlog <- function(object, ...) {
  assert_that(
    is_hermes_data(object)
  )
  DESeq2::rlog(counts(object))
} 
```

## Test
```{r}
library(hermes)
object <- HermesData(summarized_experiment)
h_vst(object)
h_rlog(object)
aa <- normalize_update(object)

assay(aa, "vst")
assay(aa, "rlog")
```

The results look good but a little bit slow.
