---
title: "Design: Add Heatmap Depicting (normalized) Counts"
author: "Alex Richardson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

## Objective

A plot method for `AnyHermesData` objects to produce following plots in default setting: 
1. Generate a heatmap of a single assay, data normalized
2. Title based on input dataset
3. How should the function behave if there are several assays?
4. Organize by assay name or other parameters?
5. Use filter/subset function to remove genes that are zero value for all samples? 
  a. Use subset method of hermes package to filter by low expression using subset/rowData
  b. Ultimate function will have arguments for "filter tech failure, low expr" etc
6. Review differential expression for relevance to heatmap taxonomy
7. Collapse by column groups in dataset, eg country, indication, etc?
8. Colored bands (annotations) above dendrogram, with key or labels - or label dendrogram?

## Idea

OK let's first try out how we can obtain an object that can be an input
for the heatmap function:

```{r}
library(hermes)
library(dplyr)
library(ComplexHeatmap)

# result <- (assays(normalize(hermes_data))[[1]])[1:20,] %>%
  # subset(a, subset = low_expression_flag, select = DISCSTUD == "N")

result <- hermes_data %>%
  normalize(methods = 'voom') %>%
  add_quality_flags()

result2 <- filter(result)

result <- result[[1]]
dim(result)

result
```

We subset from "not low expression" to remove the genes that don't show up in the samples, but we don't see the rows of full zeroes disappearing:

```{r}
notZeroVector <- rowSums(result)
notZeroVector <- notZeroVector > mean(notZeroVector)
result <- result[notZeroVector,]
```

And here is how we can then obtain the heatmap:

```{r}
Heatmap_plot <- ComplexHeatmap::Heatmap(result)
draw(Heatmap_plot)
```

## Prototype

Here is the prototype function:

```{r Prototype}

NormalizedHeatmap <- function(datasetInput)
  {
  filteredData <- datasetInput %>%
    normalize(methods = 'voom') %>%
    add_quality_flags() %>%
    assays()
  
  filteredData <- filteredData[[1]]
  
  # Example for "remove low expression" is to remove those with total expression in the sample lower than the mean, this can be edited with a threshold, or removing the zero-expression genes in the assay. The subset method with argument select = !low_expression_flag did not seem to filter anything out when applied to this example assay.
  
  
 
  Heatmap_plot <- ComplexHeatmap::Heatmap(result)
  draw(Heatmap_plot)

}

NormalizedHeatmap(hermes_data)

```


