<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference • hermes</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://insightsengineering.github.io/hermes/main/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-125641273-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-125641273-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hermes</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1.9009</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/hermes.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
<!-- start dropdown for versions --> <li class="nav-item dropdown">
        <a href="#" class="nav-link dropdown-toggle"
          data-bs-toggle="dropdown" role="button"
          aria-expanded="false" aria-haspopup="true"
          id="dropdown-versions">Versions</a> <div class="dropdown-menu" aria-labelledby="dropdown-versions"> <a class="dropdown-item" data-toggle="tooltip" title="" href="https://insightsengineering.github.io/hermes/main">main</a> </div></li>
<!-- end dropdown for versions -->
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/insightsengineering/hermes">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="package">Package<a class="anchor" aria-label="anchor" href="#package"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="hermes-package.html">hermes</a></code> <code><a href="hermes-package.html">hermes-package</a></code> 
        </dt>
        <dd><code>hermes</code> Package</dd>
      </dl></div><div class="section level2">
      <h2 id="classes">Classes<a class="anchor" aria-label="anchor" href="#classes"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="HermesData-class.html">HermesData()</a></code> <code><a href="HermesData-class.html">HermesDataFromMatrix()</a></code> 
        </dt>
        <dd><code>HermesData</code> and <code>RangedHermesData</code></dd>
      </dl><dl><dt>
          
          <code><a href="calc_cor.html">correlate(<i>&lt;AnyHermesData&gt;</i>)</a></code> <code><a href="calc_cor.html">autoplot(<i>&lt;HermesDataCor&gt;</i>)</a></code> 
        </dt>
        <dd>Correlation between Sample Counts of <code>AnyHermesData</code></dd>
      </dl><dl><dt>
          
          <code><a href="calc_pca.html">calc_pca()</a></code> 
        </dt>
        <dd>Principal Components Analysis Calculation</dd>
      </dl><dl><dt>
          
          <code><a href="top_genes.html">top_genes()</a></code> <code><a href="top_genes.html">autoplot(<i>&lt;HermesDataTopGenes&gt;</i>)</a></code> 
        </dt>
        <dd>Derivation of Top Genes</dd>
      </dl><dl><dt>
          
          <code><a href="summary.html">summary()</a></code> <code><a href="summary.html">show(<i>&lt;HermesDataSummary&gt;</i>)</a></code> 
        </dt>
        <dd>Summary Method for <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="pca_cor_samplevar.html">correlate(<i>&lt;HermesDataPca&gt;</i>)</a></code> <code><a href="pca_cor_samplevar.html">autoplot(<i>&lt;HermesDataPcaCor&gt;</i>)</a></code> 
        </dt>
        <dd>Correlation of Principal Components with Sample Variables</dd>
      </dl><dl><dt>
          
          <code><a href="diff_expression.html">diff_expression()</a></code> <code><a href="diff_expression.html">autoplot(<i>&lt;HermesDataDiffExpr&gt;</i>)</a></code> 
        </dt>
        <dd>Differential Expression Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="connect_biomart.html">connect_biomart()</a></code> 
        </dt>
        <dd>Connection to BioMart</dd>
      </dl><dl><dt>
          
          <code><a href="GeneSpec.html">GeneSpec</a></code> 
        </dt>
        <dd>R6 Class Representing a Gene (Signature) Specification</dd>
      </dl></div><div class="section level2">
      <h2 id="constructors">Constructors<a class="anchor" aria-label="anchor" href="#constructors"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="HermesData-class.html">HermesData()</a></code> <code><a href="HermesData-class.html">HermesDataFromMatrix()</a></code> 
        </dt>
        <dd><code>HermesData</code> and <code>RangedHermesData</code></dd>
      </dl><dl><dt>
          
          <code><a href="gene_spec.html">gene_spec()</a></code> 
        </dt>
        <dd><code>GeneSpec</code> Constructor</dd>
      </dl></div><div class="section level2">
      <h2 id="internal-helper-functions">Internal Helper Functions<a class="anchor" aria-label="anchor" href="#internal-helper-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="all_na.html">all_na()</a></code> 
        </dt>
        <dd>Checks Whether All Missing</dd>
      </dl><dl><dt>
          
          <code><a href="cat_with_newline.html">cat_with_newline()</a></code> 
        </dt>
        <dd>Concatenate and Print with Newline</dd>
      </dl><dl><dt>
          
          <code><a href="validate.html">validate_counts()</a></code> <code><a href="validate.html">validate_cols()</a></code> <code><a href="validate.html">validate_row_data()</a></code> <code><a href="validate.html">validate_col_data()</a></code> <code><a href="validate.html">validate_names()</a></code> <code><a href="validate.html">validate_prefix()</a></code> 
        </dt>
        <dd>Internal Helper Functions for Validation of <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="h_map_pos.html">h_map_pos()</a></code> 
        </dt>
        <dd>Helper Function For Matching Map Values to Names</dd>
      </dl><dl><dt>
          
          <code><a href="normalize.html">normalize(<i>&lt;AnyHermesData&gt;</i>)</a></code> <code><a href="normalize.html">h_cpm()</a></code> <code><a href="normalize.html">h_rpkm()</a></code> <code><a href="normalize.html">h_tpm()</a></code> <code><a href="normalize.html">h_voom()</a></code> <code><a href="normalize.html">h_vst()</a></code> <code><a href="normalize.html">h_rlog()</a></code> 
        </dt>
        <dd>Normalization of <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="quality_flags.html">add_quality_flags()</a></code> <code><a href="quality_flags.html">h_low_expression_flag()</a></code> <code><a href="quality_flags.html">h_low_depth_flag()</a></code> <code><a href="quality_flags.html">h_tech_failure_flag()</a></code> <code><a href="quality_flags.html">get_tech_failure()</a></code> <code><a href="quality_flags.html">get_low_depth()</a></code> <code><a href="quality_flags.html">get_low_expression()</a></code> 
        </dt>
        <dd>Add Quality Flags</dd>
      </dl><dl><dt>
          
          <code><a href="h_diff_expr_voom.html">h_diff_expr_voom()</a></code> 
        </dt>
        <dd><code>limma</code>/<code>voom</code> Differential Expression Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="h_diff_expr_deseq2.html">h_diff_expr_deseq2()</a></code> 
        </dt>
        <dd>DESeq2 Differential Expression Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="h_pca_var_rsquared.html">h_pca_var_rsquared()</a></code> 
        </dt>
        <dd>Calculation of R2 between Sample Variable and Principal Components</dd>
      </dl><dl><dt>
          
          <code><a href="h_pca_df_r2_matrix.html">h_pca_df_r2_matrix()</a></code> 
        </dt>
        <dd>Calculation of R2 Matrix between Sample Variables and Principal Components</dd>
      </dl><dl><dt>
          
          <code><a href="h_get_annotation_biomart.html">h_get_annotation_biomart()</a></code> 
        </dt>
        <dd>Get Annotations from BioMart</dd>
      </dl><dl><dt>
          
          <code><a href="h_get_granges_by_id.html">h_get_granges_by_id()</a></code> 
        </dt>
        <dd>Conversion of BioMart Coordinates into <code>GRanges</code></dd>
      </dl><dl><dt>
          
          <code><a href="h_get_size_biomart.html">h_get_size_biomart()</a></code> 
        </dt>
        <dd>Total Length of All Exons for Genes</dd>
      </dl><dl><dt>
          
          <code><a href="h_ensembl_to_entrez_ids.html">h_ensembl_to_entrez_ids()</a></code> 
        </dt>
        <dd>Translation of Ensembl to Entrez Gene IDs</dd>
      </dl><dl><dt>
          
          <code><a href="h_strip_prefix.html">h_strip_prefix()</a></code> 
        </dt>
        <dd>Stripping Prefix from Gene IDs</dd>
      </dl><dl><dt>
          
          <code><a href="h_has_req_annotations.html">h_has_req_annotations()</a></code> 
        </dt>
        <dd>Predicate for Required Annotations</dd>
      </dl><dl><dt>
          
          <code><a href="h_short_list.html">h_short_list()</a></code> 
        </dt>
        <dd>Make a Short List of a Character Vector</dd>
      </dl><dl><dt>
          
          <code><a href="h_parens.html">h_parens()</a></code> 
        </dt>
        <dd>Parenthesize a Character Vector</dd>
      </dl><dl><dt>
          
          <code><a href="h_all_duplicated.html">h_all_duplicated()</a></code> 
        </dt>
        <dd>Finding All Duplicates in Vector</dd>
      </dl><dl><dt>
          
          <code><a href="h_unique_labels.html">h_unique_labels()</a></code> 
        </dt>
        <dd>Creation of Unique Labels</dd>
      </dl><dl><dt>
          
          <code><a href="h_df_factors_with_explicit_na.html">h_df_factors_with_explicit_na()</a></code> 
        </dt>
        <dd>Conversion to Factors with Explicit Missing Level in a <code>data.frame</code></dd>
      </dl><dl><dt>
          
          <code><a href="draw_boxplot.html">draw_boxplot()</a></code> <code><a href="draw_boxplot.html">h_draw_boxplot_df()</a></code> 
        </dt>
        <dd>Boxplot for Gene Expression Values</dd>
      </dl></div><div class="section level2">
      <h2 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="expression_set.html">expression_set</a></code> 
        </dt>
        <dd>Example <code>ExpressionSet</code> Data</dd>
      </dl><dl><dt>
          
          <code><a href="hermes_data.html">hermes_data</a></code> 
        </dt>
        <dd>Example <code>HermesData</code> Data</dd>
      </dl><dl><dt>
          
          <code><a href="summarized_experiment.html">summarized_experiment</a></code> 
        </dt>
        <dd>Example <code>SummarizedExperiment</code> Data</dd>
      </dl><dl><dt>
          
          <code><a href="multi_assay_experiment.html">multi_assay_experiment</a></code> 
        </dt>
        <dd>Example <code>MultiAssayExperiment</code> Data</dd>
      </dl></div><div class="section level2">
      <h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="annotation.html">annotation(<i>&lt;AnyHermesData&gt;</i>)</a></code> <code><a href="annotation.html">.row_data_annotation_cols</a></code> <code><a href="annotation.html">`annotation&lt;-`(<i>&lt;AnyHermesData&gt;</i>,<i>&lt;DataFrame&gt;</i>)</a></code> 
        </dt>
        <dd>Annotation Accessor and Setter</dd>
      </dl><dl><dt>
          
          <code><a href="cbind.html">cbind</a></code> 
        </dt>
        <dd>Column Binding of <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="correlate.html">correlate()</a></code> 
        </dt>
        <dd>Generic Function for Correlation Calculations</dd>
      </dl><dl><dt>
          
          <code><a href="counts.html">counts(<i>&lt;AnyHermesData&gt;</i>)</a></code> <code><a href="counts.html">`counts&lt;-`(<i>&lt;AnyHermesData&gt;</i>,<i>&lt;matrix&gt;</i>)</a></code> 
        </dt>
        <dd>Counts Accessor and Setter</dd>
      </dl><dl><dt>
          
          <code><a href="extra_data_names.html">extraColDataNames()</a></code> <code><a href="extra_data_names.html">extraRowDataNames()</a></code> 
        </dt>
        <dd>Extra Variable Names Accessor Methods</dd>
      </dl><dl><dt>
          
          <code><a href="filter.html">filter()</a></code> 
        </dt>
        <dd>Filter <code>AnyHermesData</code> on Subset Passing Default QC Flags</dd>
      </dl><dl><dt>
          
          <code><a href="genes.html">genes()</a></code> 
        </dt>
        <dd>Gene IDs Accessor</dd>
      </dl><dl><dt>
          
          <code><a href="isEmpty.html">isEmpty(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> 
        </dt>
        <dd>Checking for Empty <code>SummarizedExperiment</code></dd>
      </dl><dl><dt>
          
          <code><a href="lapply.html">lapply(<i>&lt;MultiAssayExperiment&gt;</i>)</a></code> 
        </dt>
        <dd><code>lapply</code> method for <code>MultiAssayExperiment</code></dd>
      </dl><dl><dt>
          
          <code><a href="metadata.html">metadata</a></code> 
        </dt>
        <dd>Metadata Accessor and Setter</dd>
      </dl><dl><dt>
          
          <code><a href="normalize.html">normalize(<i>&lt;AnyHermesData&gt;</i>)</a></code> <code><a href="normalize.html">h_cpm()</a></code> <code><a href="normalize.html">h_rpkm()</a></code> <code><a href="normalize.html">h_tpm()</a></code> <code><a href="normalize.html">h_voom()</a></code> <code><a href="normalize.html">h_vst()</a></code> <code><a href="normalize.html">h_rlog()</a></code> 
        </dt>
        <dd>Normalization of <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="prefix.html">prefix()</a></code> 
        </dt>
        <dd>Prefix Accessor</dd>
      </dl><dl><dt>
          
          <code><a href="query.html">query()</a></code> 
        </dt>
        <dd>Query Gene Annotations from a Connection</dd>
      </dl><dl><dt>
          
          <code><a href="rbind.html">rbind</a></code> 
        </dt>
        <dd>Row Binding of <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="rename.html">rename(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="rename.html">rename(<i>&lt;data.frame&gt;</i>)</a></code> 
        </dt>
        <dd>Renaming Contents of <code>SummarizedExperiment</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="samples.html">samples(<i>&lt;AnyHermesData&gt;</i>)</a></code> 
        </dt>
        <dd>Sample IDs Accessor</dd>
      </dl><dl><dt>
          
          <code><a href="show.html">show(<i>&lt;HermesData&gt;</i>)</a></code> <code><a href="show.html">show(<i>&lt;RangedHermesData&gt;</i>)</a></code> 
        </dt>
        <dd>Show Method for <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="subset.html">subset</a></code> 
        </dt>
        <dd>Subsetting <code>AnyHermesData</code> Objects</dd>
      </dl><dl><dt>
          
          <code><a href="summary.html">summary()</a></code> <code><a href="summary.html">show(<i>&lt;HermesDataSummary&gt;</i>)</a></code> 
        </dt>
        <dd>Summary Method for <code>AnyHermesData</code> Objects</dd>
      </dl></div><div class="section level2">
      <h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="quality_flags.html">add_quality_flags()</a></code> <code><a href="quality_flags.html">h_low_expression_flag()</a></code> <code><a href="quality_flags.html">h_low_depth_flag()</a></code> <code><a href="quality_flags.html">h_tech_failure_flag()</a></code> <code><a href="quality_flags.html">get_tech_failure()</a></code> <code><a href="quality_flags.html">get_low_depth()</a></code> <code><a href="quality_flags.html">get_low_expression()</a></code> 
        </dt>
        <dd>Add Quality Flags</dd>
      </dl><dl><dt>
          
          <code><a href="calc_pca.html">calc_pca()</a></code> 
        </dt>
        <dd>Principal Components Analysis Calculation</dd>
      </dl><dl><dt>
          
          <code><a href="diff_expression.html">diff_expression()</a></code> <code><a href="diff_expression.html">autoplot(<i>&lt;HermesDataDiffExpr&gt;</i>)</a></code> 
        </dt>
        <dd>Differential Expression Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="set_tech_failure.html">set_tech_failure()</a></code> 
        </dt>
        <dd>Set Technical Failure Flags</dd>
      </dl><dl><dt>
          
          <code><a href="top_genes.html">top_genes()</a></code> <code><a href="top_genes.html">autoplot(<i>&lt;HermesDataTopGenes&gt;</i>)</a></code> 
        </dt>
        <dd>Derivation of Top Genes</dd>
      </dl><dl><dt>
          
          <code><a href="connect_biomart.html">connect_biomart()</a></code> 
        </dt>
        <dd>Connection to BioMart</dd>
      </dl><dl><dt>
          
          <code><a href="df_cols_to_factor.html">df_cols_to_factor()</a></code> 
        </dt>
        <dd>Conversion of Eligible Columns to Factor Variables in a <code>DataFrame</code></dd>
      </dl><dl><dt>
          
          <code><a href="colPrinComp1.html">colPrinComp1()</a></code> 
        </dt>
        <dd>First Principal Component (PC1) Gene Signature</dd>
      </dl><dl><dt>
          
          <code><a href="colMeanZscores.html">colMeanZscores()</a></code> 
        </dt>
        <dd>Mean Z-score Gene Signature</dd>
      </dl><dl><dt>
          
          <code><a href="wrap_in_mae.html">wrap_in_mae()</a></code> 
        </dt>
        <dd>Wrap in MAE</dd>
      </dl><dl><dt>
          
          <code><a href="col_data_with_genes.html">col_data_with_genes()</a></code> 
        </dt>
        <dd>Sample Variables with Selected Gene Information</dd>
      </dl><dl><dt>
          
          <code><a href="inner_join_cdisc.html">inner_join_cdisc()</a></code> 
        </dt>
        <dd>Inner Joining a Genes with a CDISC Data Set</dd>
      </dl><dl><dt>
          
          <code><a href="cut_quantile.html">cut_quantile()</a></code> 
        </dt>
        <dd>Cutting a Numeric Vector into a Factor of Quantile Bins</dd>
      </dl></div><div class="section level2">
      <h2 id="graphs">Graphs<a class="anchor" aria-label="anchor" href="#graphs"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="draw_libsize_hist.html">draw_libsize_hist()</a></code> 
        </dt>
        <dd>Histogram of Library Sizes</dd>
      </dl><dl><dt>
          
          <code><a href="draw_libsize_qq.html">draw_libsize_qq()</a></code> 
        </dt>
        <dd>Q-Q Plot of Library Sizes</dd>
      </dl><dl><dt>
          
          <code><a href="draw_libsize_densities.html">draw_libsize_densities()</a></code> 
        </dt>
        <dd>Density Plot of (Log) Counts Distributions</dd>
      </dl><dl><dt>
          
          <code><a href="draw_nonzero_boxplot.html">draw_nonzero_boxplot()</a></code> 
        </dt>
        <dd>Boxplot of Non-Zero Genes</dd>
      </dl><dl><dt>
          
          <code><a href="draw_genes_barplot.html">draw_genes_barplot()</a></code> 
        </dt>
        <dd>Stacked Barplot of Low Expression Genes by Chromosome</dd>
      </dl><dl><dt>
          
          <code><a href="draw_heatmap.html">draw_heatmap()</a></code> 
        </dt>
        <dd>Heatmap for Gene Expression Counts</dd>
      </dl><dl><dt>
          
          <code><a href="draw_scatterplot.html">draw_scatterplot()</a></code> 
        </dt>
        <dd>Scatterplot for Gene Expression Values</dd>
      </dl><dl><dt>
          
          <code><a href="draw_barplot.html">draw_barplot()</a></code> 
        </dt>
        <dd>Barplot for Gene Expression Percentiles</dd>
      </dl><dl><dt>
          
          <code><a href="draw_boxplot.html">draw_boxplot()</a></code> <code><a href="draw_boxplot.html">h_draw_boxplot_df()</a></code> 
        </dt>
        <dd>Boxplot for Gene Expression Values</dd>
      </dl><dl><dt>
          
          <code><a href="plot_all.html">autoplot(<i>&lt;AnyHermesData&gt;</i>)</a></code> 
        </dt>
        <dd>All Standard Plots in Default Setting</dd>
      </dl></div><div class="section level2">
      <h2 id="control-functions">Control Functions<a class="anchor" aria-label="anchor" href="#control-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="control_normalize.html">control_normalize()</a></code> 
        </dt>
        <dd>Control Settings for Counts Normalization</dd>
      </dl><dl><dt>
          
          <code><a href="control_quality.html">control_quality()</a></code> 
        </dt>
        <dd>Control for Specified Quality Flags</dd>
      </dl></div><div class="section level2">
      <h2 id="assertions">Assertions<a class="anchor" aria-label="anchor" href="#assertions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="assertion_arguments.html">assertion_arguments</a></code> 
        </dt>
        <dd>Standard Assertion Arguments</dd>
      </dl><dl><dt>
          
          <code><a href="check_proportion.html">check_proportion()</a></code> <code><a href="check_proportion.html">assert_proportion()</a></code> <code><a href="check_proportion.html">test_proportion()</a></code> <code><a href="check_proportion.html">expect_proportion()</a></code> 
        </dt>
        <dd>Check for proportion</dd>
      </dl><dl><dt>
          
          <code><a href="assertions.html">is_class()</a></code> <code><a href="assertions.html">is_hermes_data()</a></code> <code><a href="assertions.html">is_counts_vector()</a></code> <code><a href="assertions.html">is_list_with()</a></code> <code><a href="assertions.html">one_provided()</a></code> <code><a href="assertions.html">is_constant()</a></code> 
        </dt>
        <dd>Additional Assertions for <code>assert_that</code></dd>
      </dl></div><div class="section level2">
      <h2 id="constants">Constants<a class="anchor" aria-label="anchor" href="#constants"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="annotation.html">annotation(<i>&lt;AnyHermesData&gt;</i>)</a></code> <code><a href="annotation.html">.row_data_annotation_cols</a></code> <code><a href="annotation.html">`annotation&lt;-`(<i>&lt;AnyHermesData&gt;</i>,<i>&lt;DataFrame&gt;</i>)</a></code> 
        </dt>
        <dd>Annotation Accessor and Setter</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Daniel Sabanés Bové, Namrata Bhatia, Stefanie Bienert, Benoit Falquet, Haocheng Li, Jeff Luong, Lyndsee Midori Zhang, Alex Richardson, Simona Rossomanno, Tim Treis, Mark Yan, Naomi Chang, Chendi Liao, Carolyn Zhang, Joseph N. Paulson, F. Hoffmann-La Roche AG.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

